### PR TITLE
Member CRUD

### PR DESCRIPTION
**Pull request by nurbisak**

### CODE REVIEW

File: src/main/java/kg/peaksoft/taskTrackerb6/api/MemberApi.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import org.springframework.data.repository.query.Param;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("api/members")
+@RequiredArgsConstructor
+@PreAuthorize("hasAnyAuthority('ADMIN')")
```
Comment (datka-mcoder) [inline src/main/java/kg/peaksoft/taskTrackerb6/api/MemberApi.java:19]: remove

File: src/main/java/kg/peaksoft/taskTrackerb6/api/MemberApi.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+@CrossOrigin(origins = "*", maxAge = 3600)
+@Tag(name = "Member API", description = "Member endpoints for Admin")
```
Comment (datka-mcoder) [inline src/main/java/kg/peaksoft/taskTrackerb6/api/MemberApi.java:21]: These methods are not only for the admin, user can also use them

File: src/main/java/kg/peaksoft/taskTrackerb6/db/service/MemberService.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+import kg.peaksoft.taskTrackerb6.exceptions.NotFoundException;
+import lombok.RequiredArgsConstructor;
+import org.hibernate.jdbc.Work;
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@Service
+@RequiredArgsConstructor
```
Comment (datka-mcoder) [inline src/main/java/kg/peaksoft/taskTrackerb6/db/service/MemberService.java:19]: add @Transactional

File: src/main/java/kg/peaksoft/taskTrackerb6/dto/response/AllMemberResponse.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.Setter;
+
+import java.util.List;
+
+@Getter
+@Setter
+//@AllArgsConstructor
```
Comment (datka-mcoder) [inline src/main/java/kg/peaksoft/taskTrackerb6/dto/response/AllMemberResponse.java:11]: remove

File: src/main/java/kg/peaksoft/taskTrackerb6/db/service/MemberService.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+public class MemberService {
+
+    private final UserRepository userRepository;
+    private final WorkspaceRepository workspaceRepository;
+    private final CardRepository cardRepository;
+    private final BoardRepository boardRepository;
+
+    public List<MemberResponse> searchByEmailOrName(Long id, String emailOrName) {
+        Workspace workspace = workspaceRepository.findById(id).orElseThrow(
+                () -> new NotFoundException("Work space not found"));
```
Comment (nurbekovn) [inline src/main/java/kg/peaksoft/taskTrackerb6/db/service/MemberService.java:29]: Write together :  Workspace.

File: src/main/java/kg/peaksoft/taskTrackerb6/db/service/MemberService.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        return userRepository.searchByEmailOrName(emailOrName,workspace.getId());
+    }
+
+    public AllMemberResponse getAllMembers(Long id) {
+        Card card = cardRepository.findById(id).orElseThrow(
+                () -> new NotFoundException("Card not found"));
```
Comment (nurbekovn) [inline src/main/java/kg/peaksoft/taskTrackerb6/db/service/MemberService.java:35]: Add :  Card with ID not found !

File: src/main/java/kg/peaksoft/taskTrackerb6/db/service/MemberService.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        Board board = boardRepository.findById(card.getBoard().getId()).orElseThrow(
+                () -> new NotFoundException("Board not found"));
```
Comment (nurbekovn) [inline src/main/java/kg/peaksoft/taskTrackerb6/db/service/MemberService.java:37]: Here too . Use  ID , please !

File: src/main/java/kg/peaksoft/taskTrackerb6/db/service/MemberService.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        Workspace workspace = workspaceRepository.findById(card.getWorkspace().getId()).orElseThrow(
+                () -> new NotFoundException("Workspace not found"));
```
Comment (nurbekovn) [inline src/main/java/kg/peaksoft/taskTrackerb6/db/service/MemberService.java:39]: Here too .

File: src/main/java/kg/peaksoft/taskTrackerb6/db/service/MemberService.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        AllMemberResponse members = new AllMemberResponse();
+        List<MemberResponse> boardMembers = new ArrayList<>();
+        List<MemberResponse> workspaceMembers = new ArrayList<>();
+        for(User user : board.getMembers()) {
+            boardMembers.add(new MemberResponse(user));
+        }
+        for(UserWorkSpace workSpace : workspace.getUserWorkSpaces()) {
+            User user = userRepository.findUserByWorkSpaceId(workSpace.getId()).orElseThrow(
+                    () -> new NotFoundException("User not found"));
```
Comment (nurbekovn) [inline src/main/java/kg/peaksoft/taskTrackerb6/db/service/MemberService.java:48]: Here too.

File: src/main/java/kg/peaksoft/taskTrackerb6/db/service/MemberService.java [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+        }
+        members.setBoardMembers(boardMembers);
+        members.setWorkspaceMembers(workspaceMembers);
+        return members;
+    }
+
+
+    public MemberResponse assignMemberToCard(Long memberId, Long cardId) {
+        User user = userRepository.findById(memberId).orElseThrow(
+                () -> new NotFoundException("User not found"));
```
Comment (nurbekovn) [inline src/main/java/kg/peaksoft/taskTrackerb6/db/service/MemberService.java:59]: Here too.

File: src/main/java/kg/peaksoft/taskTrackerb6/db/service/MemberService.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        Card card = cardRepository.findById(cardId).orElseThrow(
+                () -> new NotFoundException("Card not found"));
```
Comment (nurbekovn) [inline src/main/java/kg/peaksoft/taskTrackerb6/db/service/MemberService.java:61]: Here too.

File: src/main/java/kg/peaksoft/taskTrackerb6/api/MemberApi.java [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+import lombok.RequiredArgsConstructor;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("api/members")
+@RequiredArgsConstructor
+@CrossOrigin(origins = "*", maxAge = 3600)
+@Tag(name = "Member API")
```
Comment (i-Coderr) [inline src/main/java/kg/peaksoft/taskTrackerb6/api/MemberApi.java:17]: add description

**Review by i-Coderr — COMMENTED** [review]
Well done!

**Pull request merged**